sound/soc/sof/ipc4-pcm.c

```diff
@@ -469,10 +469,57 @@ static int sof_ipc4_pcm_close(struct snd_soc_component *component,
 	return 0;
 }
 
+static int sof_ipc4_get_time_info(struct snd_soc_component *component,
+				  struct snd_pcm_substream *substream,
+				  struct timespec64 *system_ts,
+				  struct timespec64 *audio_ts,
+				  struct snd_pcm_audio_tstamp_config *audio_tstamp_config,
+				  struct snd_pcm_audio_tstamp_report *audio_tstamp_report)
+{
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct snd_sof_pcm *spcm;
+
+	spcm = snd_sof_find_spcm_dai(component, rtd);
+	if (!spcm)
+		return -EINVAL;
+
+	if (substream->runtime->hw.info & SNDRV_PCM_INFO_HAS_LINK_ATIME &&
+	    audio_tstamp_config->type_requested == SNDRV_PCM_AUDIO_TSTAMP_TYPE_LINK) {
+		u64 nsec;
+
+		/* timecounter is not ready, waiting for it */
+		if (!spcm->stream[substream->stream].accuracy) {
+			msleep(2 * SOF_IPC4_WCLK_CALIBRATION_DELAY_MS);
```

Comment:
Hmm, this is very dangerous. get_time_info() is called from interrupt context (snd_pcm_update_hw_ptr0)) and calling msleep() here does not seem correct.

Could we instead provide an zero timestamp until calibration is completed? Not ideal either, but I don't think we can just sleep.

Reply:
why snd_pcm_update_hw_ptr0 is called from interrupt context ? For SOF , it is called from a work thread : snd_sof_pcm_period_elapsed_work. Is it called by other cases ?

Reply:
in theory we don't need the work thread for period elapsed. IIRC this was a work-around added by @keyonjie to prevent a stop IPC from being sent while we are already processing an IPC.

I am not aware of any other implementation that uses a work queue. If we find a better solution we should get rid of the work queue.

Reply:
thanks, I removed the sleep